core/trino-main/src/test/java/io/trino/operator/output/TestPositionsAppender.java

```diff
@@ -0,0 +1,349 @@
+
+    private DictionaryBlock dictionaryBlock(Block dictionary, int[] ids)
+    {
+        return new DictionaryBlock(0, ids.length, dictionary, ids, false, randomDictionaryId());
+    }
+
+    private DictionaryBlock dictionaryBlock(Type type, int positionCount, int dictionarySize, float nullRate)
+    {
+        Block dictionary = createRandomBlockForType(type, dictionarySize, nullRate);
+        return createRandomDictionaryBlock(dictionary, positionCount);
+    }
+
+    private RunLengthEncodedBlock rleBlock(Block value, int positionCount)
+    {
+        return new RunLengthEncodedBlock(value, positionCount);
+    }
+
+    private RunLengthEncodedBlock rleBlock(Type type, int positionCount)
+    {
+        Block rleValue = createRandomBlockForType(type, 1, 0);
+        return new RunLengthEncodedBlock(rleValue, positionCount);
+    }
+
+    private RunLengthEncodedBlock nullRleBlock(Type type, int positionCount)
+    {
+        Block rleValue = nullBlock(type, 1);
+        return new RunLengthEncodedBlock(rleValue, positionCount);
+    }
+
+    private Block partiallyNullBlock(Type type, int positionCount)
+    {
+        return createRandomBlockForType(type, positionCount, 0.5F);
+    }
+
+    private Block notNullBlock(Type type, int positionCount)
+    {
+        return createRandomBlockForType(type, positionCount, 0);
+    }
+
+    private Block nullBlock(Type type, int positionCount)
+    {
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, positionCount);
+        for (int i = 0; i < positionCount; i++) {
+            blockBuilder.appendNull();
+        }
+        return blockBuilder.build();
+    }
+
+    private Block emptyBlock(Type type)
+    {
+        return type.createBlockBuilder(null, 0).build();
+    }
+
+    private void testNullRle(Type type, Block source)
+    {
+        PositionsAppender positionsAppender = POSITIONS_APPENDER_FACTORY.create(type, 10, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+
+        positionsAppender.append(new IntArrayList(IntStream.range(0, source.getPositionCount()).toArray()), source);
+        Block actual = positionsAppender.build();
+        assertTrue(actual.isNull(0));
+        assertEquals(actual.getPositionCount(), 2);
+        assertTrue(actual instanceof RunLengthEncodedBlock, actual.toString());
+    }
+
+    private void testAppend(Type type, List<BlockView> inputs)
+    {
+        testAppend(type, inputs, (positionsAppender, input) -> positionsAppender.append(input.getPositions(), input.getBlock()));
+    }
+
+    private void testAppend(Type type, List<BlockView> inputs, BiConsumer<PositionsAppender, BlockView> appendCall)
+    {
+        PositionsAppender positionsAppender = POSITIONS_APPENDER_FACTORY.create(type, 10, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+        long initialRetainedSize = positionsAppender.getRetainedSizeInBytes();
+
+        inputs.forEach(input -> appendCall.accept(positionsAppender, input));
+        long sizeInBytes = positionsAppender.getSizeInBytes();
```

Comment:
you can make sure that retained size is >= size in bytes

Reply:
good point and it actually caught a bug in `SlicePositionAppender`